File: WEB programming/programming-www/w4code14/Stack_class.rb
# Stack2_class.rb - a class to implement a stack-like 
#                   structure in an array
class Stack2_class

# Constructor - parameter is the size of the stack - default is 100
  def initialize(len = 100)
    @stack_ref = Array.new(len)
    @max_len = len
    @top_index = -1
  end

# push method
  def push(number)
    if @top_index == @max_len
      puts "Error in push - stack is full"
    else
      @top_index += 1
      @stack_ref[@top_index] = number
    end
  end

# pop method
  def pop()
    if @top_index == -1
      puts "Error in pop - stack is empty"
    else
      @top_index -= 1
    end
  end

# top method
  def top()
    if @top_index > -1
      return @stack_ref[@top_index]
    else
      puts "Error in top - no elements"
    end
  end

# top2 method
  def top2
    if @top_index > 0
      return @stack_ref[@top_index - 1]
    else
      puts "Error in top2 - there are not 2 elements"
    end
  end

# empty method
  def empty()
    @topIndex == -1
  end

end

File: WEB programming/programming-www/w4code12/stackClass.cs
// stackClass.cs - a file containing two class definitions
//  1. A class, StackClass, that implements a stack in an 
//     array of length 100
//  2. A class, TstStack, to test the StackClass

using System;

public class StackClass {
  private int [] stackRef;
  private int maxLen,
              topIndex;

// A constructor for StackClass
  public StackClass() {  
    stackRef = new int [100]; 
    maxLen = 99;
    topIndex = -1;
  }

// The push method
  public void push(int number) {
    if (topIndex == maxLen)
      Console.WriteLine("Error in push—stack is full");
    else stackRef[++topIndex] = number;
  }

// The pop method
  public void pop() {
    if (topIndex == -1)
      Console.WriteLine("Error in pop—stack is empty");
    else --topIndex;
  }

// The top method
  public int top() {return (stackRef[topIndex]);}

// The empty predicate method
  public bool empty() {return (topIndex == -1);}
}

// A class to test StackClass 
public class TstStack {
  public static void Main() {

// Create an instance of StackClass
    StackClass myStack = new StackClass();

// Test push, top, pop, and empty
    myStack.push(42);
    myStack.push(29);
    Console.WriteLine("29 is: " + myStack.top());
    myStack.pop();
    Console.WriteLine("42 is: " + myStack.top());
    myStack.pop();
    myStack.pop();  // Produces an error message
  }
}
